### PR TITLE
Added CI steps to install and run SWIG to generate wrappers.

### PR DESCRIPTION
**Pull request by iamAzeem**
Added CI steps to install and run SWIG to generate wrappers.

Tested after commenting `SKIP_SWIG: true` and by adding`touch ext/*/*.i`.
Kindly adjust the environment variable accordingly.

Signed-off-by: Azeem Sajid <azeem.sajid@gmail.com>

### CODE REVIEW

**Comment by iamAzeem** [comment]
@robinst:
Kindly review, test, and merge this.
The environment variable wasn't working for me.
I didn't spend much time on it though.
Kindly verify that on your side.
I'm not sure if that's needed anymore.
Maybe, you have some use-case for that.

Regards!

**Comment by robinst** [comment]
> Tested after commenting `SKIP_SWIG: true` and by adding `touch ext/*/*.i`

Can you remove the `SKIP_SWIG` variable from the workflow and add the `touch` step? Then it looks good to merge.

**Comment by iamAzeem** [comment]
@robinst: It's done.

Regarding #90, a couple of CentOS instances with GCC and Clang can also be set up in the workflow.
That needs some detailed testing in isolation though. I'll check it out later.

**Comment by Sigill** [comment]
@iamAzeem Even if it only takes an extra minute to build Swig, you should try to cache it.

Here is how I did it on another project: 
https://github.com/Sigill/cppcheck-vcs-utils/blob/82b14ac2f92f88c179b7b5124d49b349992d1db2/.github/workflows/python-package.yml#L35

**Comment by iamAzeem** [comment]
@Sigill: Thank you for your suggestion!

Please take a look at this cache workflow:
https://github.com/iamAzeem/taglib-ruby/blob/2487f334175ba6ec37cb8d83a3a38dbbdc2b49dc/.github/workflows/ci.yml#L25-L33

Here's its latest run:
https://github.com/iamAzeem/taglib-ruby/actions/runs/505468812

IIRC, the jobs run in parallel. There are only two Ruby versions here.
Only one would get to cache SWIG. That's what I'm observing from the logs.
I'm not sure if caching makes much sense here.

Can you please verify if it's being done correctly in my fork?
Let me know if there's some issue with the configuration.

Regards!

**Comment by Sigill** [comment]
I believe what you've done doesn't work (disclaimer: I'm no github-actions-cache expert, I've just used it once successfully and thought it was nice).

I'll explain why, but first, let's clarify why caching makes sense even if there's only one build flavor.

In you case, you have two jobs, that run indeed in parallel in will more or less start at the same time.
In the first run, there's nothing in the cache, so both jobs will build Swig and put it in the cache (the 2nd job will overwrite the cache entry created at the end of the 1st job).
All the next run will however be able to retrieve something from the cache, which is nice when you iterate on a pull request.

You workflow should not work because it:

1. Clones the sources and perform the build in `./swig`.
2. Installs it to the default install prefix, which is either `/usr` or `/usr/local`.
3. Caches `./swig`.

Any subsequent run will:

1. Retrieve ./swig from the cache.
2. Skip the build step as the cache has been hit.
3. ❗ Fail to execute Swig since `/usr{/local}/bin/swig` has not been restored from the cache.

What you want to do is cache the result of the install step (because it's the minimal set of files we need).
Currently, you can't because you are installing swig to `/usr{/local}`.
This is the reason why in my example I've specified [a custom install prefix](https://github.com/Sigill/cppcheck-vcs-utils/blob/82b14ac2f92f88c179b7b5124d49b349992d1db2/.github/workflows/python-package.yml#L45).
You should be able to do the same with `./configure --prefix=../swig` (you might need an absolute path).
⚠️ Since the swig executable will not be in the PATH, you'll have to do it manually ([example](https://github.com/Sigill/cppcheck-vcs-utils/blob/82b14ac2f92f88c179b7b5124d49b349992d1db2/.github/workflows/python-package.yml#L56)).

A last note before you update the workflow:
The cache key you specified is now used, and there's no way to change its content (well, you can remove the `if: ...cache-hit` directive, but it makes using the cache useless).
Basically, that key is now unusable for new content (for the next week, it will then be automatically discarded).
Add versioning to the key (e.g. `${{ runner.os }}-swig-3.0.7-1` and increment the last number each time you change anything related to building or caching Swig in the workflow).

**Comment by iamAzeem** [comment]
@Sigill: I followed your advice and tried what you said earlier but no luck. :(

Can you please take a look at this [branch](https://github.com/iamAzeem/taglib-ruby/tree/ci-workflow-run-swig) of my fork and test it on your side?
Maybe, you'll spot the issue right away.

In addition, I believe it's good to have the caching feature for Bundler as suggested here:
https://github.com/actions/cache/blob/main/examples.md#ruby---bundler

In fact, other steps such as TagLib installation may also benefit from this.
In my last commit here, I have incorporated the custom `--prefix` so the caching may be configured accordingly.

Regards!

**Comment by Sigill** [comment]
@iamAzeem You were close, it took me also quite some time to get it right.

The main issue was trying to use an environment variable to define the path of a cache entry `path: ~/$SWIG_DIR`.
The `$` syntax is only available in `run` sections (because it is basically a shell, so environment variables are substituted as usual).

For other parts of a workflow file, [according to the docs](https://docs.github.com/en/actions/reference/environment-variables#determining-when-to-use-default-environment-variables-or-contexts):

> [...] you must use the env context to read environment variables; this is because workflow keys (such as if) require the variable to be substituted during workflow processing before it is sent to the runner.

`path: ~/${{ env.SWIG_DIR }}`

Have a look at <https://github.com/Sigill/taglib-ruby/blob/ci-workflow-run-swig-2/.github/workflows/ci.yml>.
I've been able to cache taglib too. The build is now down to 45s \o/.

A few other things: 
- I had to add back `gem install bundler` otherwise the `bundle` command would not be available when using Ruby 2.4.
  Not really sure why.
- Since there's currently no `Gemfile.lock` I would not advise caching the bundle directory.
  And even though, this step only takes 6-7s, using a cache might reduce it to 3s, so...
- `runner.os` is "Linux". For cache purpose, especially when targeting multiple OS, using the distribution name+version might be better.
- And even when using the `matrix.os` trick, I don't really advise the `ubuntu-latest`.
  When the actual version changes (which is actually supposed to happen [right now](https://github.com/actions/virtual-environments/issues/1816), you might end-up retrieving from the cache something that doesn't match the system it has been built-for (or having the CI suddenly breaking for no apparent reason).

**Comment by iamAzeem** [comment]
@Sigill: That's progress! :) Actually, I did try `path` with `env.SWIG_DIR` but maybe there was some other issue with another part (s) of the configuration that it didn't work.

I just checked the `Gemfile.lock` on my machine and here are its contents:

```
$ cat Gemfile.lock 
PATH
  remote: .
  specs:
    taglib-ruby (1.1.0)

GEM
  remote: http://rubygems.org/
  specs:
    kramdown (2.3.0)
      rexml
    power_assert (2.0.0)
    rake (13.0.3)
    rake-compiler (0.9.9)
      rake
    rexml (3.2.4)
    shoulda-context (1.2.2)
    test-unit (3.3.9)
      power_assert
    yard (0.9.26)

PLATFORMS
  ruby

DEPENDENCIES
  bundler (>= 1.2, < 3)
  kramdown (~> 2.3.0)
  rake-compiler (~> 0.9)
  shoulda-context (~> 1.0)
  taglib-ruby!
  test-unit (~> 3.1)
  yard (~> 0.9.26)

BUNDLED WITH
   2.1.4
```

I'm not sure why we cannot add this and cache the bundled gems.
Maybe, @robinst would be able to shed some light on this.

---

Apart from that, I ran Rubocop and got this summary:

```
76 files inspected, 1356 offenses detected
```

I believe we can add the linting part as well while we're at it with Rubocop pinned to a specific version.
Reference: https://docs.github.com/en/actions/guides/building-and-testing-ruby#linting-your-code

**Comment by iamAzeem** [comment]
As of now, the test coverage is **33.76%**.

Here's the coverage summary after `bundle exec rake test`:
```
Coverage report generated for Unit Tests to ~/taglib-ruby/coverage. 525 / 1555 LOC (33.76%) covered.
```

PFA the coverage report with Simplecov: [coverage.zip](https://github.com/robinst/taglib-ruby/files/5872593/coverage.zip)

**Comment by robinst** [comment]
> I'm not sure why we cannot add this and cache the bundled gems.
> Maybe, @robinst would be able to shed some light on this.

Sorry I'm a bit out of the loop with Ruby development. Yeah I don't see why not? Try it and we'll see if it works or run into problems.

> I believe we can add the linting part as well while we're at it with Rubocop pinned to a specific version.

Yeah, but let's do it in a separate PR and merge this one first. Is this one ready to merge?

**Comment by iamAzeem** [comment]
@robinst: Sure. A separate PR makes sense.
@Sigill: Would you like something here?

**Comment by Sigill** [comment]
This PR looks good to me, thank you @iamAzeem.

**Comment by robinst** [comment]
Merged, thanks @iamAzeem!
